# Atividade prática 2.1

Considere o conjunto de dados `PimaIndiansDiabetes`, disponível na biblioteca `mlbench`, utilizado em estudos de saúde e epidemiologia para análise de fatores associados ao diabetes _mellitus_ tipo 2. Esta base, em particular, contém informações clínicas e laboratoriais de mulheres adultas, incluindo variáveis numéricas relacionadas a medidas fisiológicas e um desfecho binário que indica a presença ou ausência de diabetes. Diante dessas características, e utilizando o ambiente R, responda aos itens a seguir referentes à primeira etapa da AED, correspondente à visualização da estrutura e dos tipos de dados.

In [1]:
import pandas as pd

### 1. Carregue o conjunto de dados `PimaIndiansDiabetes` e identifique:
    - O número total de observações e de variáveis.
    - Os nomes das variáveis e sua organização na base de dados.
    - Indique quais funções do R (ou python rs) são mais adequadas para essa inspeção estrutural inicial.

In [10]:
# Carregar o conjunto de dados
df = pd.read_csv("../datasets/diabetes.csv")

In [11]:
# nome das variáveis e organização na base de dados
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [12]:
df.head(n=6)

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,6,148,72,35,0,33.6,0.627,50,1
1,1,85,66,29,0,26.6,0.351,31,0
2,8,183,64,0,0,23.3,0.672,32,1
3,1,89,66,23,94,28.1,0.167,21,0
4,0,137,40,35,168,43.1,2.288,33,1
5,5,116,74,0,0,25.6,0.201,30,0


In [20]:
df.tail()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
763,10,101,76,48,180,32.9,0.171,63,0
764,2,122,70,27,0,36.8,0.340,27,0
765,5,121,72,23,112,26.2,0.245,30,0
766,1,126,60,0,0,30.1,0.349,47,1
767,1,93,70,31,0,30.4,0.315,23,0


In [13]:
df.describe()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
count,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000
mean,3.845052,120.894531,69.105469,20.536458,79.799479,31.992578,0.471876,33.240885,0.348958
std,3.369578,31.972618,19.355807,15.952218,115.244002,7.884160,0.331329,11.760232,0.476951
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.078000,21.000000,0.000000
25%,1.000000,99.000000,62.000000,0.000000,0.000000,27.300000,0.243750,24.000000,0.000000
50%,3.000000,117.000000,72.000000,23.000000,30.500000,32.000000,0.372500,29.000000,0.000000
75%,6.000000,140.250000,80.000000,32.000000,127.250000,36.600000,0.626250,41.000000,1.000000
max,17.000000,199.000000,122.000000,99.000000,846.000000,67.100000,2.420000,81.000000,1.000000


In [18]:
def glimpse(df):
    print(f"Rows: {df.shape[0]}")
    print(f"Columns: {df.shape[1]}")
    for col in df.columns:
        print(f"$ {col} <{df[col].dtype}> {df[col].head().values}")

In [19]:
glimpse(df)

Rows: 768
Columns: 9
$ Pregnancies <int64> [6 1 8 1 0]
$ Glucose <int64> [148  85 183  89 137]
$ BloodPressure <int64> [72 66 64 66 40]
$ SkinThickness <int64> [35 29  0 23 35]
$ Insulin <int64> [  0   0   0  94 168]
$ BMI <float64> [33.6 26.6 23.3 28.1 43.1]
$ DiabetesPedigreeFunction <float64> [0.627 0.351 0.672 0.167 2.288]
$ Age <int64> [50 31 32 21 33]
$ Outcome <int64> [1 0 1 0 1]


**RESPOSTAS**: 
1. Neste dataset, existem _9 variáveis_ e _768_ pessoas.

2. Segue a tabela com o nome das variáveis e sua organização. Todas possuem a mesma quantidade de instruções, logo, apenas

| **Nome variáveis**       	| **Tipo**  	|
|--------------------------	|-----------	|
| Pregnancies              	| int       	|
| Glucose                  	| int       	|
| BloodPresure             	| int       	|
| SkinThickness            	| int       	|
| Insulin                  	| int       	|
| BMI                      	| int       	|
| DiabetesPedigreeFunction 	| float     	|
| Age                      	| float     	|
| Outcome                  	| int       	|

3. As melhores funções para essa análise inicial dos dados, seriam, para cada linguagem:

| **R**     	| **Python** 	|
|-----------	|------------	|
| str()     	| info()     	|
| head()    	| head()     	|
| dim()     	| describe() 	|

### 2. Examine os tipos de dados associados a cada variável (por exemplo, numérico, inteiro, fator). Com base nessa inspeção:
    - Classifique as variáveis em contínuas, discretas e categóricas
    - Identifique o papel de cada variável no contexto de um estudo em saúde (por exemplo, variável resposta, covariável clínica ou fator de risco).



**RESPOSTA**: Para referência e ajuda na classificação das variáveis, foi utilizado o site da [Prefeitura do Paraná](https://www.saude.pr.gov.br/Pagina/Diabetes-diabetes-mellitus) sobre a Diabetes Tipo 2


| **Nome variáveis**       	| **Tipo**  	| **Classificação** 	| **Tipo associado** 	| **Range**        	|
|--------------------------	|-----------	|-------------------	|--------------------	|------------------	|
| Pregnancies              	| int       	| Discreta          	| Covariável clínica 	| 0 < x < 17 (?)   	|
| Glucose                  	| int       	| Discreta          	| Fator de risco     	| 0 < x < 199      	|
| BloodPresure             	| int       	| Discreta          	| Covariável clínica 	| 0 < x < 122      	|
| SkinThickness            	| int       	| Discreta          	| Covariável clínica 	| 0 < x < 99       	|
| Insulin                  	| int       	| Contínua          	| Fator de risco     	| 0 < x < 846      	|
| BMI                      	| int       	| Discreta          	| Fator de risco     	| 0(?) < x < 67    	|
| DiabetesPedigreeFunction 	| float     	| Discreta          	| Fator de risco     	| 0.078 < x < 2.42 	|
| Age                      	| float     	| Discreta          	| Covariável clínica 	| 21 < x < 81      	|
| Outcome                  	| int       	| Booleana          	| Variável resposta  	| 0 e 1            	|

tabela 2.1: análise mais aprofundada dos dados obtidos

### 3. Verifique a presença de valores ausentes ou de valores que possam indicar codificações especiais (por exemplo, zeros em variáveis fisiológicas que não admitem esse valor).

In [17]:
df.isna().sum() # sem valores nulos

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

**RESPOSTA**

Como pode ser visto, não há nenhum valor ausente - isso não significa que não há valores nulos, como 0. 
Contudo, há sim valores que são um pouco displicentes. Na tabela 2.1, percebe-se na coluna _Range_ que as seguintes variáveis apresentam: 
- `Pregnancies`: uma mulher teve 17 gravidezes. Apesar de não ser impossível, é incomum - a depender da época e região que o estudo foi feito 
- `Glucose`: segundo o [Ministério da Saúde](https://linhasdecuidado.saude.gov.br/portal/diabetes-mellitus-tipo-2-(DM2)-no-adulto/sou-paciente) e considerando que a unidade de medida é $mg/dl$, uma pessoa normal tem $100mg/dl$. Contudo, ter um valor 0 significa que essa pessoa está muito abaixo do normal ou, ainda por cima, não tem nenhuma glicose no corpo (mesmo que ainda seja a principal fonte de energia das células do corpo humano). Não é impossível, mas muito improvável.
- `BloodPressure`: segundo, novamente, o [Ministério da Saúde](https://www.gov.br/pt-br/noticias/saude-e-vigilancia-sanitaria/2021/05/saiba-como-descobrir-e-prevenir-a-hipertensao-arterial) e considerando a unidade de médidade de $hhMg$, uma pessoa saudável chega a 120mmHg. Logo, é um caso semelhante a variável anterior, onde existe uma pessoa com pressão arterial de 0.
- `Insulina`: sendo uma célula responsável pela redução da manutenção da glicose no sangue, faz sentido que haja medições com 0 (ainda mais para pessoas que desenvolveram diabetes)
- `BMI`: O IMC é uma fórmula representada por $\frac{peso}{altura²}$, existe registros iguais a 0. Logo, alguma pessoa há uma altura ou peso igual a 0. Dado impossível.

### 4. Utilize funções de inspeção estrutural para:
    - Obter um resumo geral da base.
    - Identificar possíveis variáveis que necessitem de transformação de tipo (por exemplo, conversão de variável resposta para fator). 
    - Discuta por que essa etapa é fundamental antes da sumarização estatística ou da visualização gráfica propriamente dita.

In [21]:
# Em python, a função que melhor descreve os dados de uma forma geral já forá utilizada anteriormente
df.describe()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
count,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000
mean,3.845052,120.894531,69.105469,20.536458,79.799479,31.992578,0.471876,33.240885,0.348958
std,3.369578,31.972618,19.355807,15.952218,115.244002,7.884160,0.331329,11.760232,0.476951
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.078000,21.000000,0.000000
25%,1.000000,99.000000,62.000000,0.000000,0.000000,27.300000,0.243750,24.000000,0.000000
50%,3.000000,117.000000,72.000000,23.000000,30.500000,32.000000,0.372500,29.000000,0.000000
75%,6.000000,140.250000,80.000000,32.000000,127.250000,36.600000,0.626250,41.000000,1.000000
max,17.000000,199.000000,122.000000,99.000000,846.000000,67.100000,2.420000,81.000000,1.000000


**RESPOSTA**

A variável que precisa ser ajustada, a priori, é a de `outcome` que, nada mais é, do que nossa variável resposta. Pela análise preliminar, ela só varia entre dois valores: 0 e 1, sendo para casos em que não há diabetes tipo 2 e casos em que há, respectivamente. A transformação dela seria para um tipo `booleano`, onde 0 seria False e 1, True.

A importância dessa etapa está no formato do _workflow_ do aprendizado de máquina. A primeira etapa é entender como estão dispostos os dados: sua organização, valores, ausências, incongruências e afins, isso tudo para que possamos aplicar um modelo congruente aos nossos dados - pois afinal, **devemos ajustar o modelo aos dados, não o contrário**. Garantir uma boa análise de como se encontra nossa base de dados é, também, uma forma de saber qual é o melhor caminho a se seguir.